src/vs/workbench/parts/files/electron-browser/views/explorerViewer.ts

```diff
@@ -949,43 +949,49 @@ export class FileDragAndDrop extends SimpleFileResourceDragAndDrop {
 
 			return updateConfirmSettingsPromise.then(() => {
 				if (res.confirmed) {
-					return TPromise.join(sources.map(source => this.doHandleExplorerDrop(tree, source, target, isCopy))).then(() => void 0);
+					const rootDropPromise = this.doHandleRootDrop(sources.filter(s => s.isRoot), target);
+					return TPromise.join(sources.filter(s => !s.isRoot).map(source => this.doHandleExplorerDrop(tree, source, target, isCopy)).concat(rootDropPromise)).then(() => void 0);
 				}
 
 				return TPromise.as(void 0);
 			});
 		});
 	}
 
-	private doHandleExplorerDrop(tree: ITree, source: FileStat, target: FileStat | Model, isCopy: boolean): TPromise<void> {
-		return tree.expand(target).then(() => {
-			if (source.isRoot) {
-				const folders = this.contextService.getWorkspace().folders;
-				let sourceIndex: number;
-				let targetIndex: number;
-				const workspaceCreationData: IWorkspaceFolderCreationData[] = [];
-				const targetUri = target instanceof FileStat ? target.resource : folders[folders.length - 1].uri;
-
-				for (let index = 0; index < folders.length; index++) {
-					if (folders[index].uri.toString() === source.resource.toString()) {
-						sourceIndex = index;
-					}
-					if (folders[index].uri.toString() === targetUri.toString()) {
-						targetIndex = index;
-					}
-					workspaceCreationData.push({
-						name: folders[index].name,
-						uri: folders[index].uri
-					});
-				}
+	private doHandleRootDrop(roots: FileStat[], target: FileStat | Model): TPromise<void> {
+		const folders = this.contextService.getWorkspace().folders;
+		let targetIndex: number;
+		const workspaceCreationData: IWorkspaceFolderCreationData[] = [];
+		const targetUri = target instanceof FileStat ? target.resource : folders[folders.length - 1].uri;
+		const rootsToMove: IWorkspaceFolderCreationData[] = [];
+
+		for (let index = 0; index < folders.length; index++) {
+			const data = {
+				name: folders[index].name,
+				uri: folders[index].uri
```

Comment:
@isidorn also careful here, you are introducing a `` property that previously might not have been there

Reply:
@bpasero how can I check if the name was there previously?

Reply:
@isidorn you don't need to, just leave `name` empty